daffodil-test/src/test/scala/org/apache/daffodil/layers/CheckDigit.scala

```diff
@@ -17,63 +17,23 @@
 
 package org.apache.daffodil.layers
 
-import org.apache.commons.io.IOUtils
-import org.apache.daffodil.schema.annotation.props.gen.LayerLengthKind
-import org.apache.daffodil.schema.annotation.props.gen.LayerLengthUnits
-import org.apache.daffodil.util.Maybe
-import org.apache.daffodil.processors.LayerLengthEv
-import org.apache.daffodil.processors.LayerBoundaryMarkEv
-import org.apache.daffodil.processors.LayerCharsetEv
-import org.apache.daffodil.processors.parsers.PState
-import org.apache.daffodil.processors.unparsers.UState
-
-import java.io._
-import org.apache.daffodil.dpath.NodeInfo.PrimType
-import org.apache.daffodil.exceptions.Assert
-import org.apache.daffodil.processors.CharsetEvBase
-import org.apache.daffodil.processors.SequenceRuntimeData
-import org.apache.daffodil.processors.SuspendableOperation
-import org.apache.daffodil.processors.charset.BitsCharsetJava
-import org.apache.daffodil.util.ByteBufferOutputStream
 import org.apache.daffodil.util.Logger
-import org.apache.daffodil.xml.NS
-import org.apache.daffodil.xml.RefQName
 
 import java.nio.ByteBuffer
-import java.nio.charset.Charset
-
 
-final class CheckDigitExplicit(
-  srd: SequenceRuntimeData,
-  charset: Charset,
-  layerLengthEv: LayerLengthEv)
-extends LayerTransformer() {
-  Assert.invariant(charset.newDecoder().maxCharsPerByte() == 1) // is a SBCS charset
-
-  private val checkDigitVRD = {
-    val varNamespace = NS("urn:org.apache.daffodil.layers.checkDigit")
-    val varQName = RefQName(Some("cd"), "checkDigit", varNamespace).toGlobalQName
-    val vrd = srd.variableMap.getVariableRuntimeData(varQName).getOrElse {
-      srd.SDE("Variable '%s' is not defined.", varQName.toExtendedSyntax)
-    }
-    srd.schemaDefinitionUnless(vrd.primType == PrimType.Int,
-      "Variable '%s' is not of type 'xs:int'.", varQName.toExtendedSyntax)
-    vrd
-  }
 
-  private val checkDigitParamsVRD = {
-    val varNamespace = NS("urn:org.apache.daffodil.layers.checkDigit")
-    val varQName = RefQName(Some("cd"), "checkDigitParams", varNamespace).toGlobalQName
-    val vrd = srd.variableMap.getVariableRuntimeData(varQName).getOrElse {
-      srd.SDE("Variable '%s' is not defined.", varQName.toExtendedSyntax)
-    }
-    srd.schemaDefinitionUnless(vrd.primType == PrimType.String,
-      "Variable '%s' is not of type 'xs:string'.", varQName.toExtendedSyntax)
-    vrd
-  }
+final class CheckDigitExplicit(name: String,
+  infoArg: LayerRuntimeInfo)
+extends ByteBufferExplicitLengthLayerTransform[Int](
+  infoArg,
+  name,
+  variablesNamespace = "urn:org.apache.daffodil.layers.checkDigit",
+  variablesPreferredNamespacePrefix = "cd",
+  localNamesAndTypesOfVariablesToRead = Seq(("checkDigitParams", "string")),
```

Comment:
In our UDF, we use `Class` for types. Thoughts on doing something similar?

Reply:
I will investigate that, vs. this notion of using the string name of a primitive type.